### ISSUE TITLE
13.0 Account Module Migration 

### ISSUE THREAD

**Issue by gdgellatly** [OCA/OpenUpgrade#2899]
I raised an issue before about how this was incorrect.

@pedrobaeza implemented a fix, but for a client database it has gone from 12 hours to 9 days to run. It seems the loop runs forever when there are large numbers of invoice lines per invoice. Just a placeholder while investigating, and any ideas welcome.

**Comment by pedrobaeza** [comment]
The new algorithm performs a loop that iterates at maximum at the number of repeated lines found in the same invoice, and the main change from the original @MiquelRForgeFlow's approach is that saves data for performing queries in the most possible efficient way. Each iteration of the loop should take the same or less time than the previous one. If previously you had 12 hours, we have to think that this is the iteration time (which is a lot by the way, and I suppose you are not talking exclusively about such query), so if you have an invoice with 10 equal lines, then the time will be <= 10 x 12 hours.

Any other thing happening means an inconsistent DB or flaw in the algorithm. This last one seems improbable, as the steps are simple, but you can check queries in the log to see.

**Comment by gdgellatly** [comment]
@pedrobaeza indeed you are correct. It previously took ~20 minutes, and now takes ~7 days. Sorry I was looking at overall log stats. The query doesn't actually take progressively less time as the long query is the partition one. In this case the longest invoice is a little over 3000 lines.

However, in seeing the data at the end, it didn't actually work anyway. The lines are still wrong. Anyway, I've got an algorithm in test, it will take a couple of days to know if it has worked and what is wrong with it (it only gets 97% coverage so need to work that out, might just be rounding). I probably won't make a PR for it, as it will be quite situation specific (no stock, single tax, tight controls etc), but will make it public and link here once confirmed working.

But actually IRT to invoices this customers data is really simple and consistent, no inventory stuff to worry about, no edits/changes/analytics.

**Comment by gdgellatly** [comment]
OK so I'm finding a few issues now in various places and will make PR's as appropriate in next few days once testing complete. 

For this particular part we match quantity=quantity however the 2 fields can have different precision. I'm not sure how we best approach although for me ROUND(val, 2) works OK.

**Comment by pedrobaeza** [comment]
Weird... please do the PR with `ROUND(val, 4)` for better tolerance.

**Comment by gdgellatly** [comment]
@pedrobaeza it is the differing precisions between the fields that is the problem. The problem here is invoice quantity precision is in this case 3, but account quantity precision is if not 2, effectively hard coded to 2 (I seem to recall this in another install debugging anglosaxon issues) so 4 won't work. Bearing in mind this is coming from v8.

**Comment by pedrobaeza** [comment]
OK, please add the comment on the code when doing the PR, and better if you put links to that places.

**Comment by gdgellatly** [comment]
Thanks, btw, this is only in the WHERE query. The values will be the actual values. Because it is post migration I hope I can do a more generalised check and set a precision appropriately.

**Comment by gdgellatly** [comment]
FYI some new issues coming up in current algorithm

1. we never write the name field, however at least in v8 multiline descriptions only took first line over. (for fun, in this database, where there are multiline descriptions, the first line is a blank string).
2. tax_line_id IS NULL - this is not always correct (again at least in v8, Zero Rated lines had tax_line_id on the line)

**Comment by gdgellatly** [comment]
In the end I took roughly the same approach as Odoo with 1 key exception, I do not allow matching values to be different whereas enterprise migration takes a close enough is good enough approach. In a database of roughly ~1m invoice lines - no stock, few products, often repeated and manually priced, multi company, multi currency within companies, multi country. Zero grouped lines.

- Only 48 invoice lines do not match - manual cleanup - generally these are where the journal is edited after invoice validation.
- There were a few duplicate matches - however this is overcome by not copying sequence.
- Runtime dropped from 7 days -> 20mins.
- By comparison the existing looping algorithm had ~350,000 incorrect move lines post migration.

Other findings (these may only relate to v8, but also any historical data coming from there)
- Account Move name field historically limited to 64 chars.
- Account move name field will not take anything after and including a new line character from invoice line.
- An invoice line will save whitespace at end, a move line name will not.
- Rounding precision may differ between invoice line qty and aml qty.
- Tax line id is rarely NULL on invoice lines.
- Beyond the 48 errors, there were 20 others where amount_untaxed was incorrect on the invoice and carried ove but manual reconciliation showed it was the amount wrong, not the line migration.

In general I think with the current algorithm, not matching on amounts in an early pass is an error. Firstly it improves accuracy, but second it gives greater coverage. A journal line amount must bear some relation to the invoice price subtotal except in grouped scenarios. Either debit/credit or amount_currency must match.

Anyway I post the code here, it is messy, and tailored with deadlines and all. It will have more extensive QA over the next week. Maybe it starts discussion, maybe not.

https://github.com/odoonz/OpenUpgrade/blob/7a8d56f989b90e56558d38bbde95aad344e5a474/addons/account/migrations/13.0.1.1/post-migration.py#L229

**Comment by pedrobaeza** [comment]
@gdgellatly please take a look to #2973

**Comment by gdgellatly** [comment]
Hi @pedrobaeza I don't even really understand it anymore tbh. The query syntax is beyond my static review capabilities. I think the only reliable way for something when it gets this complicated is a validation query. Basically prove that it worked. That's what I ended up doing. 

I have long since migrated that database using the fork above utilising roughly Odoo's approach of just an ever diminishing sieve. Although they have 150 of them, I only used 8. But even then, I have an enterprise migration at the moment where Odoo's approach fails miserably and I end up doing a whole load of post migration scripts to resolve. It is just an all over ugly problem.
